pkg/registry/registry_docker.go

```diff
@@ -0,0 +1,185 @@
+package registry
+
+import (
+	"bufio"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+
+	"github.com/docker/docker/api/types"
+	"github.com/giantswarm/microerror"
+
+	"github.com/giantswarm/retagger/pkg/images"
+)
+
+func (r *Registry) Login() error {
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("logging in to %s registry", r.host))
+
+	authConfig := types.AuthConfig{
+		Username:      r.username,
+		Password:      r.password,
+		ServerAddress: r.host,
+	}
+
+	// We don't care for the AuthenticateOKBody.IdentityToken, as it's always nil.
+	err := IsDockerLoginFailed(r.docker.RegistryLogin(context.Background(), authConfig))
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) PullImage(image string, sha string) error {
+	if image == "" {
+		return microerror.Maskf(invalidArgumentError, "image should not be empty")
+	}
+	if sha == "" {
+		return microerror.Maskf(invalidArgumentError, "%s SHA should not be empty", image)
+	}
+
+	shaName := images.ShaName(image, sha)
+
+	r.logger.Log("level", "debug", "message", fmt.Sprintf("executing: docker pull %s", shaName))
+
+	res, err := r.docker.ImagePull(context.Background(), shaName, types.ImagePullOptions{})
+	if err != nil {
+		return microerror.Maskf(err, "could not pull image")
+	}
+	defer res.Close()
+	err = r.logDocker(res)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) TagSha(sourceImage, sha, destinationImage, destinationTag string) (string, error) {
+	if sourceImage == "" {
```

Comment:
I thought about this and might have lead you in the wrong direction here:
Does it really make sense to check for empty values here? Is this something we should be doing?
I am not against this in this case here, just wondering as this is inconsistent across packages here now.

Not blocking.

Reply:
Had the option to just delete the checking or extend it to all functions, so I decided the second

Reply:
I think its generally better than getting awkward errors from docker/client or docker cli because some fields are empty and _things happen_:tm: